pulser-core/pulser/sampler/samples.py

```diff
@@ -83,37 +76,91 @@ class ChannelSamples:
     amp: np.ndarray
     det: np.ndarray
     phase: np.ndarray
-
-    slots: list[_TimeSlot] = field(default_factory=list)
+    slots: list[_TargetSlot] = field(default_factory=list)
 
     def __post_init__(self) -> None:
         assert len(self.amp) == len(self.det) == len(self.phase)
+        self.duration = len(self.amp)
 
         for t in self.slots:
             assert t.ti < t.tf  # well ordered slots
-        for t1, t2 in _pairwise(self.slots):
+        for t1, t2 in zip(self.slots, self.slots[1:]):
             assert t1.tf <= t2.ti  # no overlaps on a given channel
 
+    def extend_duration(self, new_duration: int) -> ChannelSamples:
+        """Extends the duration of the samples.
+
+        Args:
+            new_duration: The new duration for the samples (in ns).
+                Must be greater than or equal to the current duration.
+
+        Returns:
+            The extend channel samples.
+        """
+        extension = new_duration - len(self.amp)
+        if new_duration < self.duration:
+            raise ValueError("Can't extend samples to a lower duration.")
+
+        new_amp = np.pad(self.amp, (0, extension))
+        new_detuning = np.pad(self.det, (0, extension))
+        new_phase = np.pad(
+            self.phase,
+            (0, extension),
+            mode="edge" if self.phase.size > 0 else "constant",
+        )
+        return ChannelSamples(new_amp, new_detuning, new_phase, self.slots)
+
+    def modulate(self, channel_obj: Channel) -> ChannelSamples:
+        """Modulates the samples for a given channel.
+
+        Args:
+            channel_obj: The channel object for which to modulate the samples.
+
+        Returns:
+            The modulated channel samples.
+        """
+        if not isinstance(channel_obj, Channel):
+            raise TypeError("'channel_obj' must be a Channel instance.")
+
+        new_amp = channel_obj.modulate(self.amp)
+        new_detuning = channel_obj.modulate(self.det)
+        new_phase = channel_obj.modulate(self.phase, keep_ends=True)
+        return ChannelSamples(new_amp, new_detuning, new_phase, self.slots)
+
 
 @dataclass
 class SequenceSamples:
     """Gather samples of a sequence with useful info."""
 
     channels: list[str]
-    channel_samples: dict[str, ChannelSamples]
-    _addrs: dict[str, str]
-    _bases: dict[str, str]
-    _duration: int
+    samples_list: list[ChannelSamples]
+    _ch_objs: dict[str, Channel]
+
+    @property
+    def channel_samples(self) -> dict[str, ChannelSamples]:
+        """Mapping between the channel name and its samples."""
+        return dict(zip(self.channels, self.samples_list))
+
+    @property
+    def duration(self) -> int:
```

Comment:
```suggestion
    def maximum_duration(self) -> int:
```
otherwise I expect the sequence duration because of the name